File: SalmonTrollingAssistant/Domain/Services/OfflineDataManager.swift
import Foundation
import Combine

/**
 * Manages offline data access, caching strategies, and data freshness tracking
 */
class OfflineDataManager: ObservableObject {
    // Singleton instance
    static let shared = OfflineDataManager()
    
    // UserDefaults for storing offline settings
    private let userDefaults = UserDefaults.standard
    
    // Keys for offline settings
    private let offlineModeEnabledKey = "offline_mode_enabled"
    private let lastSyncTimestampKey = "last_sync_timestamp"
    private let syncStatusKey = "sync_status"
    private let dataFreshnessThresholdKey = "data_freshness_threshold"
    private let wifiOnlySyncKey = "wifi_only_sync"
    private let autoSyncEnabledKey = "auto_sync_enabled"
    
    // Publishers
    @Published var offlineModeEnabled = false
    private let syncStatusSubject = CurrentValueSubject<SyncStatus, Never>(.synced)
    private let lastSyncTimestampSubject = CurrentValueSubject<Date?, Never>(nil)
    
    // Sync status values
    enum SyncStatus: String {
        case synced = "SYNCED"
        case syncing = "SYNCING"
        case failed = "FAILED"
        case pending = "PENDING"
    }
    
    private init() {
        // Load initial values
        loadSettings()
    }
    
    // MARK: - Public API
    
    /// Publisher for sync status
    var syncStatus: AnyPublisher<SyncStatus, Never> {
        return syncStatusSubject.eraseToAnyPublisher()
    }
    
    /// Publisher for last sync timestamp
    var lastSyncTimestamp: AnyPublisher<Date?, Never> {
        return lastSyncTimestampSubject.eraseToAnyPublisher()
    }
    
    /// Check if offline mode is currently enabled
    func isOfflineModeEnabled() -> Bool {
        return offlineModeEnabled
    }
    
    /// Enable or disable offline mode
    func setOfflineMode(enabled: Bool) {
        userDefaults.set(enabled, forKey: offlineModeEnabledKey)
        offlineModeEnabled = enabled
    }
    
    /// Get the timestamp of the last successful data synchronization
    func getLastSyncTimestamp() -> Date? {
        return lastSyncTimestampSubject.value
    }
    
    /// Update the last sync timestamp to the current time
    func updateLastSyncTimestamp() {
        let now = Date()
        userDefaults.set(now.timeIntervalSince1970, forKey: lastSyncTimestampKey)
        lastSyncTimestampSubject.send(now)
    }
    
    /// Get the current sync status
    func getSyncStatus() -> SyncStatus {
        return syncStatusSubject.value
    }
    
    /// Update the sync status
    func updateSyncStatus(_ status: SyncStatus) {
        userDefaults.set(status.rawValue, forKey: syncStatusKey)
        syncStatusSubject.send(status)
    }
    
    /// Get the data freshness threshold in seconds
    /// Data older than this threshold will be considered stale
    func getDataFreshnessThreshold() -> TimeInterval {
        return userDefaults.double(forKey: dataFreshnessThresholdKey).isZero ? 
            (24 * 60 * 60) : userDefaults.double(forKey: dataFreshnessThresholdKey)
    }
    
    /// Set the data freshness threshold in seconds
    func setDataFreshnessThreshold(_ thresholdSeconds: TimeInterval) {
        userDefaults.set(thresholdSeconds, forKey: dataFreshnessThresholdKey)
    }
    
    /// Check if WiFi-only sync is enabled
    func isWifiOnlySyncEnabled() -> Bool {
        return userDefaults.bool(forKey: wifiOnlySyncKey)
    }
    
    /// Enable or disable WiFi-only sync
    func setWifiOnlySync(enabled: Bool) {
        userDefaults.set(enabled, forKey: wifiOnlySyncKey)
    }
    
    /// Check if auto-sync is enabled
    func isAutoSyncEnabled() -> Bool {
        return userDefaults.bool(forKey: autoSyncEnabledKey)
    }
    
    /// Enable or disable auto-sync
    func setAutoSync(enabled: Bool) {
        userDefaults.set(enabled, forKey: autoSyncEnabledKey)
    }
    
    /// Check if data for a specific location is fresh (not stale)
    func isDataFresh(for location: Location) -> Bool {
        // Simplified implementation for demo
        return true
    }
    
    /// Get data freshness percentage for a location (100% = completely fresh, 0% = completely stale)
    func getDataFreshnessPercentage(for location: Location) -> Int {
        // Simplified implementation for demo
        return 85
    }
    
    /// Get the timestamp when data for a location will become stale
    func getDataExpirationTime(for location: Location) -> Date? {
        // Simplified implementation for demo
        return Date().addingTimeInterval(3600) // 1 hour from now
    }
    
    /// Check if there is any cached data available for a location
    func hasCachedData(for location: Location) -> Bool {
        // Simplified implementation for demo
        return true
    }
    
    /// Clear all cached data
    func clearAllCachedData() {
        // Simplified implementation for demo
        print("Clearing all cached data")
    }
    
    /// Clear cached data for a specific location
    func clearCachedData(for location: Location) {
        // Simplified implementation for demo
        print("Clearing cached data for location: \(location.name)")
    }
    
    // MARK: - Private Methods
    
    private func loadSettings() {
        // Load offline mode setting
        offlineModeEnabled = userDefaults.bool(forKey: offlineModeEnabledKey)
        
        // Load sync status
        if let statusString = userDefaults.string(forKey: syncStatusKey),
           let status = SyncStatus(rawValue: statusString) {
            syncStatusSubject.send(status)
        } else {
            syncStatusSubject.send(.synced)
        }
        
        // Load last sync timestamp
        if let timestamp = userDefaults.object(forKey: lastSyncTimestampKey) as? TimeInterval {
            lastSyncTimestampSubject.send(Date(timeIntervalSince1970: timestamp))
        } else {
            lastSyncTimestampSubject.send(nil)
        }
        
        // Set defaults if not already set
        if userDefaults.object(forKey: dataFreshnessThresholdKey) == nil {
            userDefaults.set(24 * 60 * 60, forKey: dataFreshnessThresholdKey) // 24 hours
        }
        
        if userDefaults.object(forKey: wifiOnlySyncKey) == nil {
            userDefaults.set(true, forKey: wifiOnlySyncKey)
        }
        
        if userDefaults.object(forKey: autoSyncEnabledKey) == nil {
            userDefaults.set(true, forKey: autoSyncEnabledKey)
        }
    }
}

File: android/SalmonTrollingAssistant/app/src/main/java/com/example/salmontrollingassistant/data/service/OfflineDataManager.kt
package com.example.salmontrollingassistant.data.service

import android.content.Context
import androidx.datastore.core.DataStore
import androidx.datastore.preferences.core.Preferences
import androidx.datastore.preferences.core.booleanPreferencesKey
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.longPreferencesKey
import androidx.datastore.preferences.core.stringPreferencesKey
import androidx.datastore.preferences.preferencesDataStore
import com.example.salmontrollingassistant.data.db.LocationDao
import com.example.salmontrollingassistant.data.db.TideDao
import com.example.salmontrollingassistant.data.db.WeatherDao
import com.example.salmontrollingassistant.domain.model.Location
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.map
import java.util.Date
import javax.inject.Inject
import javax.inject.Singleton

private val Context.offlineDataStore: DataStore<Preferences> by preferencesDataStore(name = "offline_data_settings")

/**
 * Manages offline data access, caching strategies, and data freshness tracking
 */
@Singleton
class OfflineDataManager @Inject constructor(
    private val context: Context,
    private val weatherDao: WeatherDao,
    private val tideDao: TideDao,
    private val locationDao: LocationDao
) {
    // Keys for offline settings
    private val OFFLINE_MODE_ENABLED = booleanPreferencesKey("offline_mode_enabled")
    private val LAST_SYNC_TIMESTAMP = longPreferencesKey("last_sync_timestamp")
    private val SYNC_STATUS = stringPreferencesKey("sync_status")
    private val DATA_FRESHNESS_THRESHOLD = longPreferencesKey("data_freshness_threshold")
    private val WIFI_ONLY_SYNC = booleanPreferencesKey("wifi_only_sync")
    private val AUTO_SYNC_ENABLED = booleanPreferencesKey("auto_sync_enabled")
    
    // Sync status values
    enum class SyncStatus {
        SYNCED, SYNCING, FAILED, PENDING
    }
    
    /**
     * Check if offline mode is currently enabled
     */
    fun isOfflineModeEnabled(): Flow<Boolean> {
        return context.offlineDataStore.data.map { preferences ->
            preferences[OFFLINE_MODE_ENABLED] ?: false
        }
    }
    
    /**
     * Enable or disable offline mode
     */
    suspend fun setOfflineMode(enabled: Boolean) {
        context.offlineDataStore.edit { preferences ->
            preferences[OFFLINE_MODE_ENABLED] = enabled
        }
    }
    
    /**
     * Get the timestamp of the last successful data synchronization
     */
    fun getLastSyncTimestamp(): Flow<Long> {
        return context.offlineDataStore.data.map { preferences ->
            preferences[LAST_SYNC_TIMESTAMP] ?: 0L
        }
    }
    
    /**
     * Update the last sync timestamp to the current time
     */
    suspend fun updateLastSyncTimestamp() {
        context.offlineDataStore.edit { preferences ->
            preferences[LAST_SYNC_TIMESTAMP] = System.currentTimeMillis()
        }
    }
    
    /**
     * Get the current sync status
     */
    fun getSyncStatus(): Flow<SyncStatus> {
        return context.offlineDataStore.data.map { preferences ->
            val statusString = preferences[SYNC_STATUS] ?: SyncStatus.SYNCED.name
            try {
                SyncStatus.valueOf(statusString)
            } catch (e: Exception) {
                SyncStatus.FAILED
            }
        }
    }
    
    /**
     * Update the sync status
     */
    suspend fun updateSyncStatus(status: SyncStatus) {
        context.offlineDataStore.edit { preferences ->
            preferences[SYNC_STATUS] = status.name
        }
    }
    
    /**
     * Get the data freshness threshold in milliseconds
     * Data older than this threshold will be considered stale
     */
    fun getDataFreshnessThreshold(): Flow<Long> {
        return context.offlineDataStore.data.map { preferences ->
            // Default to 24 hours
            preferences[DATA_FRESHNESS_THRESHOLD] ?: (24 * 60 * 60 * 1000L)
        }
    }
    
    /**
     * Set the data freshness threshold in milliseconds
     */
    suspend fun setDataFreshnessThreshold(thresholdMs: Long) {
        context.offlineDataStore.edit { preferences ->
            preferences[DATA_FRESHNESS_THRESHOLD] = thresholdMs
        }
    }
    
    /**
     * Check if WiFi-only sync is enabled
     */
    fun isWifiOnlySyncEnabled(): Flow<Boolean> {
        return context.offlineDataStore.data.map { preferences ->
            preferences[WIFI_ONLY_SYNC] ?: true
        }
    }
    
    /**
     * Enable or disable WiFi-only sync
     */
    suspend fun setWifiOnlySync(enabled: Boolean) {
        context.offlineDataStore.edit { preferences ->
            preferences[WIFI_ONLY_SYNC] = enabled
        }
    }
    
    /**
     * Check if auto-sync is enabled
     */
    fun isAutoSyncEnabled(): Flow<Boolean> {
        return context.offlineDataStore.data.map { preferences ->
            preferences[AUTO_SYNC_ENABLED] ?: true
        }
    }
    
    /**
     * Enable or disable auto-sync
     */
    suspend fun setAutoSync(enabled: Boolean) {
        context.offlineDataStore.edit { preferences ->
            preferences[AUTO_SYNC_ENABLED] = enabled
        }
    }
    
    /**
     * Check if data for a specific location is fresh (not stale)
     */
    suspend fun isDataFresh(locationId: String): Boolean {
        val threshold = getDataFreshnessThreshold().first()
        val currentTime = System.currentTimeMillis()
        
        // Check weather data freshness
        val weatherData = weatherDao.getCurrentWeather(locationId)
        val weatherFresh = weatherData != null && 
                          (currentTime - weatherData.cacheTimestamp) < threshold
        
        // Check tide data freshness
        val tideData = tideDao.getCurrentTide(locationId)
        val tideFresh = tideData != null && 
                       (currentTime - tideData.cacheTimestamp) < threshold
        
        return weatherFresh && tideFresh
    }
    
    /**
     * Get data freshness percentage for a location (100% = completely fresh, 0% = completely stale)
     */
    suspend fun getDataFreshnessPercentage(locationId: String): Int {
        val threshold = getDataFreshnessThreshold().first()
        val currentTime = System.currentTimeMillis()
        
        // Check weather data freshness
        val weatherData = weatherDao.getCurrentWeather(locationId)
        val weatherAge = if (weatherData != null) {
            currentTime - weatherData.cacheTimestamp
        } else {
            threshold
        }
        
        // Check tide data freshness
        val tideData = tideDao.getCurrentTide(locationId)
        val tideAge = if (tideData != null) {
            currentTime - tideData.cacheTimestamp
        } else {
            threshold
        }
        
        // Calculate average freshness
        val weatherFreshness = ((threshold - weatherAge).coerceAtLeast(0) * 100 / threshold).toInt()
        val tideFreshness = ((threshold - tideAge).coerceAtLeast(0) * 100 / threshold).toInt()
        
        return (weatherFreshness + tideFreshness) / 2
    }
    
    /**
     * Get the timestamp when data for a location will become stale
     */
    suspend fun getDataExpirationTime(locationId: String): Date? {
        val threshold = getDataFreshnessThreshold().first()
        
        // Get the most recent cache timestamp between weather and tide data
        val weatherData = weatherDao.getCurrentWeather(locationId)
        val tideData = tideDao.getCurrentTide(locationId)
        
        val weatherTimestamp = weatherData?.cacheTimestamp ?: 0L
        val tideTimestamp = tideData?.cacheTimestamp ?: 0L
        
        val mostRecentTimestamp = maxOf(weatherTimestamp, tideTimestamp)
        
        return if (mostRecentTimestamp > 0) {
            Date(mostRecentTimestamp + threshold)
        } else {
            null
        }
    }
    
    /**
     * Check if there is any cached data available for a location
     */
    suspend fun hasCachedData(locationId: String): Boolean {
        val weatherData = weatherDao.getCurrentWeather(locationId)
        val tideData = tideDao.getCurrentTide(locationId)
        
        return weatherData != null || tideData != null
    }
    
    /**
     * Clear all cached data
     */
    suspend fun clearAllCachedData() {
        weatherDao.clearAllWeatherData()
        tideDao.clearAllTideData()
    }
    
    /**
     * Clear cached data for a specific location
     */
    suspend fun clearCachedData(locationId: String) {
        weatherDao.deleteCurrentWeather(locationId)
        weatherDao.deleteForecast(locationId)
        tideDao.deleteCurrentTide(locationId)
        tideDao.deleteTideForecast(locationId)
    }
    
    /**
     * Get all locations that have cached data
     */
    suspend fun getLocationsWithCachedData(): List<Location> {
        val weatherLocations = weatherDao.getLocationsWithWeatherData()
        val tideLocations = tideDao.getLocationsWithTideData()
        
        val locationIds = (weatherLocations + tideLocations).distinct()
        return locationDao.getLocationsByIds(locationIds).map { it.toLocation() }
    }
}